File: credit.c
// Takes a credit card number as an input and outputs whether it is a valid number (using luhn´s algorithm) and if it is VISA, AMEX or MASTERCARD

#include <stdio.h>
#include <cs50.h>
#include <math.h>


int main(void)
{
    bool boolean = true;
    long long cardnumber = 0;
    int count = 0;


    // GET INPUT
    do
    {
        cardnumber = get_long_long("What is your number?");
        long long digits = cardnumber;
        while (digits > 0)
        {
            digits = digits/10;
            count++;
        }
        if (count != 13 && count != 15 && count != 16)
        {
            cardnumber = 0;
        }
   
    }
    while (cardnumber < 0);
   
    
    //LUHN´s ALGORITHM
    //GET A LIST OUT OF THE CARDNUMBER
    int numbers[count];
    
    for (int i = count - 1; i >= 0; i--)
    {
        numbers[i] = cardnumber % 10;
        cardnumber /= 10;
    }
   
   
    //DETERMINE IF THE NUMBER OF DIGITS IS ODD OR EVEN, THEN SET STARTING VAR FOR THE FOR LOOPS BELOW
    int a = 0;
    int b = 1;
    if (count % 2 != 0)
    {
        a = 1;
        b = 0;
    }
   
   
    //SUM OF THE ODD DIGITS
    int sum1 = 0;
    for (int i = a; i <=count - a; i += 2)
    {
        if (numbers[i]*2 > 9)
        {
            sum1 += 1;
            sum1 += numbers[i]*2 % 10;
        }
        else
        {
            sum1 += numbers[i]*2;
        }
        
    }
    
    
    //SUM OF THE EVEN NUMBERS
    int sum2 = 0;
    for (int i = b; i <= count - b ; i += 2)
    {
        sum2 += numbers[i];
       
    }
   
    
    
    //HAS THE SUM OF BOTH SUMs A ZERO AS THE LAST DIGIT?
    if ((sum1 + sum2) % 10 != 0)
    {
        boolean = false;
    }
    
    
    //DETERMINE WHAT KIND OF CREDIT CARD IT IS
    int two_digs = numbers[0]*10 + numbers[1];
    if (boolean != true)
    {
        printf("INVALID\n");
    }
    else if ((count == 15) && (two_digs == 34 || two_digs == 37))
    {
        printf("AMEX\n");
    }
    else if ((count == 16) && (two_digs == 51 || two_digs == 52 || two_digs == 53 || two_digs == 54 || two_digs == 55 ))
    {
        printf("MASTERCARD\n");
    }
    else if ((count == 13 || count == 16) && (numbers[0] == 4))
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
    return 0
}

File: credit.py
# Takes as an input a card number
# And determines whether the number is valid and if, what kind of card it is

import sys


def main():
    card_number = input("Your card number:")
    lenght = len(card_number)

    if (card_number.isdigit()) == False:
        print("INVALID")
        sys.exit(1)
        
    if (12 < int(card_number) > 17) == False:
        print("INVALID")
        sys.exit(1)
        
    if luhn(card_number) == False:
        print("INVALID")
        sys.exit(1)

    
    first_digits = card_number[0] + card_number[1]
    if lenght == 15 and first_digits in ["34", "37"]:
        print("AMEX")
    elif lenght == 16 and first_digits in ["51","52","53","54","55"]:
        print("MASTERCARD")
    elif (lenght == 13 or lenght == 16) and card_number[0] == "4":
        print("VISA")
    else:
        print("INVALID")


def luhn(card_number):
    card_number = list(map(int, card_number))
    sum1 = 0
    sum2 = 0
    
    # Calculate sum2 - every other digit - starting from the second-to-last digit 
    for i in range(len(card_number) - 2, -1, -2):
        # Add only to the sum only the product's digits (so if the product is a number 16 - add 1 and 6)
        if card_number[i] * 2 >= 10:
            sum1 += 1
            sum1 += card_number[i] * 2 % 10
        else:
            sum1 += card_number[i] * 2
            
    # Calculate sum2 - every other digit - starting from the last digit
    for i in range(len(card_number) - 1, -1, -2):
        sum2 += card_number[i]

    # Calculate final outcome
    if (sum1 + sum2) % 10 == 0:
        return True
    else:
        return False

main()
